independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java

```diff
@@ -0,0 +1,158 @@
+package io.quarkus.arc.impl;
+
+import io.quarkus.arc.InjectableContext;
+import io.quarkus.arc.ManagedContext;
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Singleton;
+
+/**
+ * Encapsulates all contexts used in the container.
+ * <p>
+ * The idea is to avoid unnecessary map usage for built-in contexts.
+ */
+class Contexts {
+
+    // Built-in contexts
+    final ManagedContext requestContext;
+    final InjectableContext applicationContext;
+    final InjectableContext singletonContext;
+
+    // Used to optimize the getContexts(Class<? extends Annotation>)
+    private final List<InjectableContext> applicationContextSingleton;
+    private final List<InjectableContext> singletonContextSingleton;
+    private final List<InjectableContext> requestContextSingleton;
+
+    // Lazily computed list of contexts for a scope
+    private final ClassValue<List<InjectableContext>> unoptimizedContexts;
+
+    // Precomputed values
+    final Set<Class<? extends Annotation>> scopes;
+
+    Contexts(ManagedContext requestContext, InjectableContext applicationContext,
+            InjectableContext singletonContext, Map<Class<? extends Annotation>, List<InjectableContext>> contexts) {
+        this.requestContext = requestContext;
+        this.applicationContext = applicationContext;
+        this.singletonContext = singletonContext;
+
+        this.applicationContextSingleton = Collections.singletonList(applicationContext);
+        this.singletonContextSingleton = Collections.singletonList(singletonContext);
+        // request singleton is only set if there is only one context registered for @RequestScoped (99% of cases)
```

Comment:
This comment is not true? It seems to me that `requestContextSingleton` is always set to a non-`null` value. (And that seems correct, from what I can tell. The comment is probably just stale?)

Reply:
Yes, it is :facepalm: